src/sycl/MoEPrepareInputs.cpp

```diff
@@ -0,0 +1,513 @@
+  return;
+}
+
+void shuffle_rows(const torch::Tensor& input_tensor, const torch::Tensor& dst2src_map, torch::Tensor& output_tensor) {
+  TORCH_CHECK(
+      input_tensor.scalar_type() == output_tensor.scalar_type(),
+      "Input and output tensors must have the same data type");
+  SYCL_DISPATCH_FLOATING_TYPES_AND2(
+      at::ScalarType::BFloat16, at::ScalarType::Half, input_tensor.scalar_type(), "shuffle_rows_kernel_impl", [&]() {
+        shuffle_rows_kernel_impl<scalar_t>(input_tensor, dst2src_map, output_tensor);
+      });
+  return;
+}
+
+template <typename T, typename T1>
+struct ApplyShuffleMulSum {
+  ApplyShuffleMulSum(
+      const T* input,
+      T* output,
+      const int32_t* dst2src_map,
+      const T1* factors,
+      const int32_t topk,
+      const int32_t hidden_dim,
+      const int32_t bs_hidden_dim)
+      : input_(input),
+        output_(output),
+        dst2src_map_(dst2src_map),
+        factors_(factors),
+        topk_(topk),
+        hidden_dim_(hidden_dim),
+        bs_hidden_dim_(bs_hidden_dim) {}
+
+  void operator()(sycl::nd_item<1> item) const {
+    int out_tkn_id = item.get_group(0);
+    float sum_val = 0;
+
+    for (int i = item.get_local_id(0); i < hidden_dim_; i += bs_hidden_dim_) {
+      for (int k = 0; k < topk_; ++k) {
+        int src_perm_offset = out_tkn_id * topk_ + k;
+        int src_index = dst2src_map_[src_perm_offset];
+        float src_val = static_cast<float>(input_[src_index * hidden_dim_ + i]);
+        float weight = 0;
+        if (factors_ != nullptr) {
+          weight = static_cast<float>(factors_[out_tkn_id * topk_ + k]);
+        }
+        sum_val += weight * src_val;
+      }
+      output_[out_tkn_id * hidden_dim_ + i] = sum_val;
+    }
+  }
+  const T* input_;
+  const int32_t* dst2src_map_;
+  T* output_;
+  const T1* factors_;
+  const int32_t topk_;
+  const int32_t hidden_dim_;
+  const int32_t bs_hidden_dim_;
+};
+
+template <typename T, typename T1>
+void apply_shuffle_mul_sum_impl(
+    const T* input,
+    T* output,
+    const int32_t* dst2src_map,
+    const T1* factors,
+    const uint32_t out_tkns,
+    const uint32_t out_hidden_dims,
+    const int topk) {
+  auto stream = at::xpu::getCurrentXPUStream();
+  auto queue = stream.queue();
+
+  using Kernel = ApplyShuffleMulSum<T, T1>;
+
+  auto dev_id = dpcppGetDeviceIdOfCurrentQueue();
+  uint32_t max_wg_size = dpcppMaxWorkGroupSize(dev_id);
+  uint32_t max_out_hidden_dims = static_cast<uint32_t>(sycl::min(max_wg_size, out_hidden_dims));
+
+  sycl::range<1> global_range{out_tkns * max_out_hidden_dims};
+  sycl::range<1> local_range{max_out_hidden_dims};
+
+  Kernel task(input, output, dst2src_map, factors, topk, out_hidden_dims, max_out_hidden_dims);
+
+  sycl_kernel_submit(global_range, local_range, queue, task);
+  return;
+}
+
+void apply_shuffle_mul_sum(
+    const torch::Tensor& input,
+    torch::Tensor& output,
+    const torch::Tensor& permutation,
+    const std::optional<torch::Tensor>& factors) {
```

Comment:
check `factors` should be the same dtype with input if this is always the case.

the following if-else can be saved